agb/src/interrupt.rs

```diff
@@ -21,37 +28,344 @@ pub enum Interrupt {
+    for (i, root) in unsafe { INTERRUPT_TABLE.iter().enumerate() } {
+        if (1 << i) & interrupt != 0 {
+            root.trigger_interrupts();
+        }
+    }
+
+    interrupt
+}
+
+pub struct InterruptClosureBounded<'a> {
+    c: InterruptClosure,
+    _phantom: PhantomData<&'a ()>,
+    _unpin: PhantomPinned,
+}
+
+struct InterruptClosure {
+    closure: *const (dyn Fn()),
+    next: Cell<*const InterruptClosure>,
+    root: *const InterruptRoot,
+}
+
+impl InterruptRoot {
+    fn trigger_interrupts(&self) {
+        let mut c = self.next.get();
+        while !c.is_null() {
+            let closure_ptr = unsafe { &*c }.closure;
+            let closure_ref = unsafe { &*closure_ptr };
+            closure_ref();
+            c = unsafe { &*c }.next.get();
+        }
+    }
+}
+
+impl Drop for InterruptClosure {
+    fn drop(&mut self) {
+        let root = unsafe { &*self.root };
+        root.reduce();
+        let mut c = root.next.get();
+        let own_pointer = self as *const _;
+        if c == own_pointer {
+            unsafe { &*self.root }.next.set(self.next.get());
+            return;
+        }
+        loop {
+            let p = unsafe { &*c }.next.get();
+            if p == own_pointer {
+                unsafe { &*c }.next.set(self.next.get());
+                return;
+            }
+            c = p;
+        }
+    }
+}
+
+fn interrupt_to_root(interrupt: Interrupt) -> &'static InterruptRoot {
+    unsafe { &INTERRUPT_TABLE[interrupt as usize] }
+}
+
+fn get_interrupt_handle_root<'a>(
+    f: &'a dyn Fn(),
+    root: &InterruptRoot,
+) -> InterruptClosureBounded<'a> {
+    InterruptClosureBounded {
+        c: InterruptClosure {
+            closure: unsafe { core::mem::transmute(f as *const _) },
+            next: Cell::new(core::ptr::null()),
+            root: root as *const _,
+        },
+        _phantom: PhantomData,
+        _unpin: PhantomPinned,
+    }
+}
+
+pub fn get_interrupt_handle(
+    f: &(dyn Fn() + Send + Sync),
+    interrupt: Interrupt,
+) -> InterruptClosureBounded {
+    let root = interrupt_to_root(interrupt);
+
+    get_interrupt_handle_root(f, root)
+}
+
+pub fn add_interrupt<'a>(interrupt: Pin<&'a InterruptClosureBounded<'a>>) {
+    let root = unsafe { &*interrupt.c.root };
+    root.add();
+    let mut c = root.next.get();
+    if c.is_null() {
+        root.next.set((&interrupt.c) as *const _);
+        return;
+    }
+    loop {
+        let p = unsafe { &*c }.next.get();
+        if p.is_null() {
+            unsafe { &*c }.next.set((&interrupt.c) as *const _);
+            return;
+        }
+
+        c = p;
+    }
+}
+
+#[macro_export]
+macro_rules! add_interrupt_handler {
+    ($interrupt: expr, $handler: expr) => {
+        let a = $handler;
+        let a = $crate::interrupt::get_interrupt_handle(&a, $interrupt);
+        let a = unsafe { core::pin::Pin::new_unchecked(&a) };
+        $crate::interrupt::add_interrupt(a);
+    };
+}
+
+#[test_case]
```

Comment:
feels a bit weird to have these tests here and then the mutex definition below this. Maybe better to push this to the bottom of the file?